src/webportal/src/app/user/fabric/components/CustomPassword.jsx

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import React, {useRef, useState, useEffect} from 'react';
+import {PropTypes} from 'prop-types';
+
+import {TextField, Callout, DirectionalHint, DefaultButton, getTheme} from 'office-ui-fabric-react';
+
+function CustomPassword(props) {
+  const {onChange, onFocus, onBlur, componentRef, ...otherProps} = props;
+
+  useEffect(() => {
+    if (componentRef && passwordInputRef) {
+      componentRef.current = passwordInputRef.current;
+    }
+    return () => {
+      if (componentRef) {
+        componentRef.current = null;
+      }
+    };
+  });
+
+  const passwordInputRef = useRef(null);
+  const randomPasswordTargetRef = useRef(null);
+  const randomPasswordButtonRef = useRef(null);
+
+  const [showRandomPassword, setShowRandomPassword] = useState(false);
+  const handleOnFocusPassword = (event) => {
+    if (!passwordInputRef.current.value) {
+      setShowRandomPassword(true);
+    }
+
+    if (onFocus) {
+      onChange(event);
+    }
+  };
+  const handleOnLostFocusPassword = (event) => {
+    if (!event.relatedTarget || !randomPasswordButtonRef.current
+      || event.relatedTarget.id !== randomPasswordButtonRef.current.props.id) {
+      setShowRandomPassword(false);
+    }
+
+    if (onBlur) {
+      onChange(event);
+    }
+  };
+  const handleOnChangePassword = (event, newValue) => {
+    if (event.target.value) {
+      setShowRandomPassword(false);
+    } else {
+      setShowRandomPassword(true);
+    }
+    setPassword(newValue);
+
+    if (onChange) {
+      onChange(event, newValue);
+    }
+  };
+
+  const generateRandomPassword = () => {
+    return Array(8)
+      .fill('0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz')
+      .map((x) => x[Math.floor(Math.random() * x.length)])
+      .join('');
```

Comment:
you can use `Math.random().toString(36).slice(-8)` to generate random string

Reply:
Yes, `Math.random().toString(36).slice(-8)` is a great suggestion and more simple, but here I want the random password can have up-case/lower-case/number.